contribute hklpy2 package

prjemian opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| Rust            | `@conda-forge/help-rust`      |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Gitter channel][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io

All apologies in advance if your recipe PR does not receive prompt attention.
This is a high volume repository and the reviewers are volunteers. Review times
vary depending on the number of reviewers on a given language team and may be days
or weeks. We are always looking for more staged-recipe reviewers. If you are
interested in volunteering, please contact a member of @conda-forge/core.
We'd love to have the help!
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

prjemian commented:
@conda-forge/staged-recipes : This is ready for review, when you get a chance. Thanks!

prjemian commented:
@conda-forge/help-python This is Python code.

ocefpaf commented on recipes/hklpy2/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+build:
+  script: {{ PYTHON }} -m pip install . -vv --no-deps --no-build-isolation
+  number: 0
+  skip: true  # [python_impl == 'pypy']
+  skip: true  # [py<310 or win or osx]
+
```
This looks like it can be noarch. I did not see any reason to skip Windows or mac OS.
If the reason is a missing dependency on those platforms, we still prefer noarch to safe resources. The package won't be installable there and would be ready if/when the missing dependencies are available.

```suggestion
  noarch: python

```

prjemian replied:
Beginning to understand this more.

I should add the `hkl` package but mark it as linux 64-bit only..
Also, the associated entry-point should also be linux 64-bit only.
I need to learn how to do that.

prjemian replied:
And the unit tests involving the `hkl` package need to be made Linux 86-bit only.

prjemian replied:
Linux-only for now.

ocefpaf replied:
It is OK for noarch package to fails on platforms with a missing dependency. We prefer noarch b/c it is faster to build and the package will be ready when the missing dependency is available on those platforms. TL;DR, if the reason is only b/c `hkl` is missing on other platforms, it should still be noarch.

prjemian replied:
> OK for noarch package to fails on platforms with a missing dependency

Now, I understand.  Switched to `noarch: python` and added version minimum for Python.  Builds as expected.

ocefpaf commented on recipes/hklpy2/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+
+build:
+  script: {{ PYTHON }} -m pip install . -vv --no-deps --no-build-isolation
+  number: 0
```
You have some entrypoints there. You will should probably also add some command line testing for them in the testing section below.

```suggestion
  number: 0
  entry_points:
    - no_op = hklpy2.backends.no_op:NoOpSolver
    - hkl_soleil = hklpy2.backends.hkl_soleil:HklSolver
    - th_tth = hklpy2.backends.th_tth_q:ThTthSolver
```

prjemian replied:
Good catch!

prjemian replied:
done

ocefpaf commented on recipes/hklpy2/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+    - pip
+  run:
+    - python
+    - packaging
```
This one doesn't seem to be to in the package metadata.

prjemian replied:
As the CI reported, `ophyd` needs `packaging` but that is not added with the ophyd package.  I'll contact that package and add `packaging` to this project's requirements until ophyd is revised.

prjemian replied:
It's in the ophyd source but not in the conda-forge [recipe](https://github.com/conda-forge/ophyd-feedstock/blob/b8e8016dad28515c4bf138eee26c5fe45772e68d/recipe/meta.yaml#L18-L29).

prjemian replied:
done

ocefpaf commented on recipes/hklpy2/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+
+requirements:
+  host:
+    - python
```
```suggestion
    - python >=3.10
```

prjemian replied:
In the next revision (linux only), this change will not be needed.

prjemian replied:
no need for version pin now

ocefpaf commented on recipes/hklpy2/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+    - setuptools >=64.0
+    - pip
+  run:
+    - python
```
```suggestion
    - python >=3.10
```

prjemian replied:
same, will not change

prjemian commented:
@ocefpaf : I'll post again when this is ready for next review.

prjemian commented:
@conda-forge/help-python @conda-forge/staged-recipes Ready for review again.  I've addressed all the comments above.  Still can't build as `noarch`.  This package is limited to linux-only due to dependence on the [`hkl` package](https://anaconda.org/conda-forge/hkl) that is only compiled for linnux 64-bit now.

prjemian commented:
@conda-forge/help-python Ready again for review.  Thanks for your time.

ocefpaf commented on recipes/hklpy2/meta.yaml:
```diff
@@ -0,0 +1,54 @@
+
+requirements:
+  host:
+    - "python >=3.10"
+    - setuptools-scm >=8
+    - setuptools >=64.0
+    - pip
+  run:
+    - "python >=3.10"
```
```suggestion
    - python >=3.10
    - setuptools-scm >=8
    - setuptools >=64.0
    - pip
  run:
    - python >=3.10
```

prjemian replied:
In YAML, 3.10 == 3.1 ; it's a float

ocefpaf replied:
They are not necessary here in the recipes.

prjemian replied:
Done

ocefpaf commented on recipes/hklpy2/meta.yaml:
```diff
@@ -0,0 +1,54 @@
+  imports:
+    - hklpy2
+  commands:
+    - pip check
```
You may want to add tests for the entrypoints here, something like outputting the help would do.

prjemian replied:
Since there is no command-line entry point (this package provides a library, not an application), there is no command-line help.  The entry-points are reported by this test command: https://github.com/conda-forge/staged-recipes/blob/84f167d8787fec83e751da040c19e0e789f9fc5c/recipes/hklpy2/meta.yaml#L40

This test command could be expanded by testing for the number of entry-points expected.

```bash
python -c "import hklpy2; assert len(hklpy2.solvers())==3; print(hklpy2.solvers())"
```

prjemian commented:
Upstream problem with the `hkl` package involving the path to the gobject-introspection support there.  I'll ping again when this PR is ready.

prjemian commented:
@conda-forge/help-python Ready again for review. Thanks for your time.  As expected, builds on linux_64 only at this time.
